xdsl/dialects/arith.py

```diff
@@ -1223,6 +1227,64 @@ class MinnumfOp(FloatingPointLikeBinaryOperation):
     traits = traits_def(Pure())
 
 
+@irdl_op_definition
+class BitcastOp(IRDLOperation):
+    name = "arith.bitcast"
+
+    input = operand_def(signlessIntegerLike | floatingPointLike)
+    result = result_def(signlessIntegerLike | floatingPointLike)
+
+    assembly_format = "$input attr-dict `:` type($input) `to` type($result)"
+
+    def __init__(self, in_arg: SSAValue | Operation, target_type: Attribute):
+        super().__init__(operands=[in_arg], result_types=[target_type])
+
+    def verify_(self) -> None:
+        # check if we have a ContainerType
+        in_type = self.input.type
+        res_type = self.result.type
+
+        if isinstance(in_type, ShapedType) and isinstance(res_type, ShapedType):
+            if in_type.get_shape() != res_type.get_shape():
+                raise VerifyException(
+                    "'arith.bitcast' operand and result types must have the same shape"
+                )
+
+            t1 = get_element_type_or_self(in_type)
```

Comment:
Could this check not go right at the start of the function? This would save you checking again in the scalar case at the bottom